src/Proto3/Suite/JSONPB/Class.hs

```diff
@@ -0,0 +1,193 @@
+encode opts = E.encodingToLazyByteString . toEncodingPB opts
+{-# INLINE encode #-}
+
+-- | 'Data.Aeson..eitherDecode' variant for deserializing a JSONPB value from a
+-- lazy 'LBS.ByteString'.
+eitherDecode :: FromJSONPB a => LBS.ByteString -> Either String a
+eitherDecode = eitherFormatError . A.eitherDecodeWith jsonEOF (A.iparse parseJSONPB)
+  where
+    eitherFormatError = either (Left . uncurry A.formatError) Right
+    {-# INLINE eitherFormatError #-}
+
+    -- NB: cribbed from aeson-1.1.1.0:Data.Aeson.Parser.Internal.jsonEOF, which
+    -- is not exported. It's simple, so we just inline it here. Might be worth
+    -- submitting a PR to export this.
+    jsonEOF :: Atto.Parser A.Value
+    jsonEOF = A.json <* skipSpace <* Atto.endOfInput
+      where
+        skipSpace :: Atto.Parser ()
+        skipSpace = Atto.skipWhile $ \w -> w == 0x20 || w == 0x0a || w == 0x0d || w == 0x09
+        {-# INLINE skipSpace #-}
+{-# INLINE eitherDecode #-}
+
+-- * Operator definitions
+
+-- | Construct a (field name, JSONPB-encoded value) tuple
+(.=) :: (HasDefault v, ToJSONPB v) => Text -> v -> FieldPB
+k .= v = FieldPB (isDefault v) (\opts -> E.pair k (toEncodingPB opts v))
+
+-- | 'Data.Aeson..:' variant for JSONPB; if the given key is missing from the
+-- object, or if it is present but its value is null, we produce the default
+-- protobuf value for the field type
+(.:) :: (FromJSONPB a, HasDefault a) => A.Object -> Text -> A.Parser a
+obj .: key = obj .:? key A..!= def
+  where
+    (.:?) = A.explicitParseFieldMaybe parseJSONPB
+
+-- * JSONPB rendering and parsing options
+
+data Options = Options
+  { optEmitDefaultValuedFields :: Bool
+  }
+
+defaultOptions :: Options
+defaultOptions = Options
+  { optEmitDefaultValuedFields = False
+  }
+
+-- * Helper types and functions
+
+dropNamedPfx :: Named a => Proxy a -> String -> String
+dropNamedPfx p = drop (length (nameOf p :: String))
+
+-- | An internal ~(field name, value) tuple (using 'E.pair' as the underlying
+-- tuple) which carries additional useful state.
+--
+-- Values of this type are produced by '.='
+--
+-- Values of this type are consumed by 'fieldsPB'
+--
+data FieldPB = FieldPB
+  { fieldIsDefaultValued :: Bool
+    -- ^ Whether or not this field is "default-valued"; i.e., whether the value
+    -- component of the key-value pair is the protobuf default value for its
+    -- type
+  , fieldTuple           :: Options -> A.Series
+    -- ^ Produce the JSONPB-encoded key-value tuple
+  }
+
+-- | @fieldsPB opts flds@ encodes the given JSONPB-encoded fields
+fieldsPB :: Options -> [FieldPB] -> A.Encoding
+fieldsPB opts@Options{..} fields = E.pairs (mconcat (fmap emit fields))
+  where
+    emit FieldPB{..}
+      | not optEmitDefaultValuedFields && fieldIsDefaultValued
+        = mempty
+      | otherwise
+        = fieldTuple opts
+
+namedEncoding :: forall e. (Named e, Show e) => e -> A.Encoding
+namedEncoding = E.string . dropNamedPfx (Proxy :: Proxy e) . show
```

Comment:
Would `TypeApplications` be useful instead of `Proxy`? I can't remember if you explored that or not.

Reply:
Yeah, this works if you use `TypeApplications` and `AllowAmbiguousTypes`